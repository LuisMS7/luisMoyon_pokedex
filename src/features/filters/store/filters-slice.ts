import { create } from "zustand";
import { Filters } from "@filters/types/filters.type";

type State = {
	pokemonName: string;
	setPokemonName: (pokemonName: string) => void;
	filters: Filters;
	filterPokemons: () => void;
};

export const useFilters = create<State>((set, get) => ({
	pokemonName: "",
	setPokemonName: (pokemonName: string) => {
		set({
			pokemonName: pokemonName.toLowerCase().trim(),
		});
	},
	filters: { pokemonName: "" },
	filterPokemons: () => {
		set({
			filters: { pokemonName: get().pokemonName },
		});
	},
}));
