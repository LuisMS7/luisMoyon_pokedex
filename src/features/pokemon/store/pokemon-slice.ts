import { create } from "zustand";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";

type State = {
	pokemons: PokemonListItemApiResponse[];
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => void;
	clearPokemons: () => void;
};

const usePokemonStore = create<State>((set, get) => ({
	pokemons: [],
	updatePokemons: (pokemons: PokemonListItemApiResponse[]) => {
		const currentPokemonsList = get().pokemons;
		const updatePokemonsList = [...currentPokemonsList, ...pokemons];
		set({
			pokemons: updatePokemonsList,
		});
	},
	setPokemons: (pokemons: PokemonListItemApiResponse[]) => {
		set({
			pokemons: pokemons,
		});
	},
	clearPokemons: () =>
		set({
			pokemons: [],
		}),
}));

export default usePokemonStore;
