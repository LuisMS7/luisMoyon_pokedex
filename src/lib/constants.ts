import { perfectSize } from "@utils/perfect-size";

export const TEXT_SMALL_SIZE = perfectSize(12);

export const TEXT_NORMAL_SIZE = perfectSize(14);

export const TEXT_BIG_SIZE = perfectSize(16);

export const TEXT_SUBTITLE_SIZE = perfectSize(20);

export const TEXT_TITLE_SIZE = perfectSize(24);

export const TEXT_SUPER_SIZE = perfectSize(32);

export const NORMAL_WEIGHT = "400";

export const BOLD_WEIGHT = "800";

export enum API_ENDPOINTS {
	all_pokemons = "api/v2/pokemon",
	pokemons_by_name = "api/v2/pokemon/{pokemon_name}",
}
