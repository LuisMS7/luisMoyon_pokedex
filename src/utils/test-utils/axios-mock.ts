import client from "@lib/axios-instance";
import MockAdapter from "axios-mock-adapter";

export const mockClient = new MockAdapter(client);

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const mockGetOnSuccess = (endpoint: string, mockResponse: any) => {
	mockClient.onGet(endpoint).reply(200, mockResponse);
};

export const mockGetOnFailure = (endpoint: string, statusCode: number) => {
	mockClient.onGet(endpoint).reply(statusCode);
};
