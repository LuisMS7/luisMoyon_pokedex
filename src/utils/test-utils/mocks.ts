import { Pokemon } from "@pokemon/types/pokemon.type";

export const mockPokemon: Pokemon = {
	id: 1,
	name: "mockPokemonName",
	// eslint-disable-next-line max-len
	img: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png",
	types: [{ type: { name: "normal", url: "" }, slot: 1 }],
	weight: 10,
	sprites: {
		back_default:
			// eslint-disable-next-line max-len
			"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/1.png",
		back_female: "",
		back_shiny: "",
		back_shiny_female: "",
		front_default: "",
		front_female: "",
		front_shiny: "",
		front_shiny_female: "",
	},
	moves: [{ move: { name: "normal", url: "" }, version_group_details: [] }],
};
