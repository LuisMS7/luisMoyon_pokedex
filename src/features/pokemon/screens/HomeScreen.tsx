import React, { useEffect, useState } from "react";
import { Image, Platform, SafeAreaView, StyleSheet, View } from "react-native";
import PokemonList from "@pokemon/components/PokemonList";
import { useApi } from "@hooks/useApi";
import getAllPokemons from "@pokemon/store/actions/get-all-pokemons";
import { usePokemonStore } from "@pokemon/store/pokemon-slice";
import LoadingModal from "@src/components/LoadingModal";
import FilterHeader from "@filters/components/FilterHeader";
import { useFilters } from "@filters/store/filters-slice";
import getPokemonByName from "@pokemon/store/actions/get-pokemon-by-name";
import { useSafeAreaInsets } from "react-native-safe-area-context";
import { useSnackbar } from "@hooks/useSnackbar";

const HomeScreen = () => {
	const [offset, setOffset] = useState(0);
	const { request, loading, response } = useApi(() =>
		getAllPokemons(offset, 20),
	);
	const {
		request: filterPokemons,
		loading: searchLoading,
		error: errorFilter,
		response: filteredResponse,
	} = useApi(() => getPokemonByName(filters.pokemonName));
	const updatePokemons = usePokemonStore((state) => state.updatePokemons);
	const pokemons = usePokemonStore((state) => state.pokemons);
	const setPokemons = usePokemonStore((state) => state.setPokemons);
	const clearPokemons = usePokemonStore((state) => state.clearPokemons);
	const filters = useFilters((state) => state.filters);
	const { showSnackbar } = useSnackbar();
	const { top } = useSafeAreaInsets();
	useEffect(() => {
		if (response) {
			updatePokemons(response.results);
			setOffset((offset) => offset + 20);
		}
	}, [response]);

	const getFilteredPokemons = async () => {
		if (filters.pokemonName) {
			filterPokemons();
		} else if (!filters.pokemonName && pokemons.length <= 1 && !loading) {
			clearPokemons();
			request();
		}
	};

	useEffect(() => {
		getFilteredPokemons();
	}, [filters]);

	useEffect(() => {
		if (filteredResponse) {
			setOffset(0);
			setPokemons(filteredResponse);
		}
	}, [filteredResponse]);

	useEffect(() => {
		if (errorFilter) {
			showSnackbar("Pokemon does not exist");
		}
	}, [errorFilter]);

	return (
		<View
			style={{
				...styles.container,
				paddingTop: Platform.OS === "android" ? top : 0,
			}}
		>
			<SafeAreaView style={styles.container}>
				<Image
					source={require("@assets/pokedex-banner.png")}
					style={{ height: "20%" }}
					resizeMode="center"
				/>
				<FilterHeader />
				<PokemonList
					pokemonsItems={pokemons}
					loadMorePokemons={request}
				/>
				<LoadingModal visible={loading || searchLoading} />
			</SafeAreaView>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		flex: 1,
		backgroundColor: "#fff",
		alignItems: "center",
		justifyContent: "flex-start",
	},
});

export default HomeScreen;
