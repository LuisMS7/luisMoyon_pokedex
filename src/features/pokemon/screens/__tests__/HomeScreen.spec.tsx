import "react-native";
import React from "react";
import { render, screen } from "@testing-library/react-native";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";
import { mockGetOnSuccess } from "@utils/test-utils/axios-mock";
import { API_ENDPOINTS } from "@lib/constants";
import HomeScreen from "@pokemon/screens/HomeScreen";
import { usePokemonStore } from "@pokemon/store/pokemon-slice";
import { NavigationContainer } from "@react-navigation/native";

jest.mock("@pokemon/store/pokemon-slice", () => ({
	usePokemonStore: jest.fn(),
}));

jest.mock("react-native-safe-area-context", () => ({
	useSafeAreaInsets: jest.fn(() => ({
		top: 20,
	})),
}));

describe("HomeScreen", () => {
	it("should render correctly", async () => {
		const mockPokemons: PokemonListItemApiResponse[] = [
			{
				name: "mockPokemon1",
				url: "",
			},
		];
		(usePokemonStore as unknown as jest.Mock).mockImplementation(
			(selector) =>
				selector({
					pokemons: mockPokemons,
					updatePokemons: jest.fn(),
					setPokemons: jest.fn(),
					clearPokemons: jest.fn(),
				}),
		);

		const mockPokemon = {
			id: 1,
			name: "mockPokemon1",
			sprites: {
				other: {
					dream_world: {
						front_default:
							// eslint-disable-next-line max-len
							"https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
					},
				},
			},
			types: [
				{
					type: {
						name: "grass",
						url: "https://pokeapi.co/api/v2/type/12/",
					},
				},
			],
			weight: 0,
			moves: [],
		};
		mockGetOnSuccess(API_ENDPOINTS.all_pokemons, mockPokemons);
		mockGetOnSuccess(
			API_ENDPOINTS.pokemons_by_name.replace(
				"{pokemon_name}",
				"mockPokemon1",
			),
			mockPokemon,
		);
		const { rerender } = render(<HomeScreen />, {
			wrapper: NavigationContainer,
		});
		await Promise.resolve();
		await rerender(<HomeScreen />);
		expect(screen.toJSON()).toMatchSnapshot();
	});
});
