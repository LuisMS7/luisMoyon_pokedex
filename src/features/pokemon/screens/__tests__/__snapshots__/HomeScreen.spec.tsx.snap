// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "paddingTop": 0,
    }
  }
>
  <RCTSafeAreaView
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flex": 1,
        "justifyContent": "flex-start",
      }
    }
  >
    <Image
      resizeMode="center"
      source={
        {
          "testUri": "../../../src/assets/pokedex-banner.png",
        }
      }
      style={
        {
          "height": "20%",
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "height": 89.4597481392934,
          "justifyContent": "center",
          "paddingHorizontal": 29.819916046431132,
          "paddingVertical": 74.54979011607783,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderColor": "black",
            "borderRadius": 29.819916046431132,
            "borderWidth": 1,
            "flexDirection": "row",
            "height": 89.4597481392934,
            "justifyContent": "center",
            "paddingLeft": 29.819916046431132,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "flex": 1,
              "height": "100%",
              "justifyContent": "space-around",
            }
          }
        >
          <Text
            style={
              {
                "fontSize": 17.89194962785868,
                "fontWeight": "800",
                "paddingTop": 4,
              }
            }
          >
            Search
          </Text>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            onChangeText={[Function]}
            placeholder="Pokemon name"
            style={
              {
                "flex": 1,
                "fontSize": 20.873941232501792,
                "fontWeight": "400",
              }
            }
            testID="pokemon-name-text-input"
            value=""
          />
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "black",
              "borderBottomRightRadius": 29.819916046431132,
              "borderTopRightRadius": 29.819916046431132,
              "height": "100%",
              "justifyContent": "center",
              "width": 149.09958023215566,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 29.819916046431132,
              }
            }
          >
            Filter
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "alignSelf": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          {
            "height": "70%",
          }
        }
      >
        <RCTScrollView
          ItemSeparatorComponent={[Function]}
          ListFooterComponent={[Function]}
          ListHeaderComponent={[Function]}
          data={
            [
              {
                "name": "mockPokemon1",
                "url": "",
              },
            ]
          }
          getItem={[Function]}
          getItemCount={[Function]}
          horizontal={true}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={50}
          stickyHeaderIndices={[]}
          style={
            {
              "flex": 1,
              "height": "70%",
            }
          }
          viewabilityConfigCallbackPairs={[]}
        >
          <View>
            <View
              onLayout={[Function]}
            >
              <View
                style={
                  {
                    "width": 29.819916046431132,
                  }
                }
              />
            </View>
            <View
              onFocusCapture={[Function]}
              onLayout={[Function]}
              style={
                [
                  {
                    "flexDirection": "row",
                  },
                  null,
                ]
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "borderColor": "#74CB48",
                    "borderRadius": 23.855932837144906,
                    "borderWidth": 5.963983209286226,
                    "flexDirection": "column",
                    "height": "100%",
                    "opacity": 1,
                    "width": 521.8485308125448,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#74CB48",
                      "fontSize": 29.819916046431132,
                      "padding": 29.819916046431132,
                      "textAlign": "right",
                      "width": "100%",
                    }
                  }
                >
                  #
                  0001
                </Text>
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flex": 1,
                    }
                  }
                />
                <Text
                  style={
                    {
                      "backgroundColor": "#74CB48",
                      "color": "white",
                      "fontSize": 35.78389925571736,
                      "paddingVertical": 29.819916046431132,
                      "textAlign": "center",
                    }
                  }
                >
                  Mockpokemon1
                </Text>
              </View>
            </View>
            <View
              onLayout={[Function]}
            >
              <View
                style={
                  {
                    "width": 29.819916046431132,
                  }
                }
              />
            </View>
          </View>
        </RCTScrollView>
      </View>
    </View>
    <Modal
      animationType="fade"
      hardwareAccelerated={false}
      transparent={true}
      visible={false}
    />
  </RCTSafeAreaView>
</View>
`;
